Feat: 닉네임으로 사용자 조회 API 구현

yangjiae12 opened the pull request:
### 👀 관련 이슈

<!-- 관련 이슈를 적어주세요 -->
#41 

### ✨ 작업한 내용

<!-- 작업한 내용을 적어주세요 -->
- 기존 완전 일치 기반 닉네임 검색을 부분 일치 + 유사도 정렬 방식으로 변경
- LevenshteinDistance를 기반으로 유사도 점수 계산 후 정렬
- 검색 결과는 최대 5개까지 반환하며, 자기 자신은 제외
- `UserSearchResponseDto`를 통해 userId, nickname, profileImageUrl, job 정보 제공
- profileImageKey는 S3Util을 사용해 실제 이미지 URL로 변환
- 인증 적용 전까지는 `userId`를 쿼리 파라미터로 받도록 구성

### 🌀 PR Point

<!-- 코드리뷰가 필요한 부분이 있다면 적어주세요 -->

### 🍰 참고사항

<!-- 참고할 사항이 있다면 적어주세요 -->
- 사용자 수가 적은 환경(1000명 미만)에서 Levenshtein 기반 유사도 정렬은 성능상 무리 없다고 하여 사용했습니다.

### 📷 스크린샷 또는 GIF

| 기능 | 스크린샷 |
| :--: | :------: |
| 닉네임으로 사용자 조회 API | <img width="1450" height="1191" alt="image" src="https://github.com/user-attachments/assets/2a9727a3-bb0f-4a0c-ab2e-e166ea3cf325" /> |
#### 조회 결과 (ja로 검색)
```json
{
  "isSuccess": true,
  "code": "USER2000",
  "message": "사용자 조회 성공",
  "result": [
    {
      "userId": 104,
      "nickname": "jay",
      "profileImageUrl": "https://mock-bucket.s3.amazonaws.com/profile/jay.jpg",
      "job": "기획자"
    },
    {
      "userId": 110,
      "nickname": "jack",
      "profileImageUrl": "https://mock-bucket.s3.amazonaws.com/profile/jack.jpg",
      "job": "학생"
    },
    {
      "userId": 103,
      "nickname": "jason",
      "profileImageUrl": "https://mock-bucket.s3.amazonaws.com/profile/jason.jpg",
      "job": "디자이너"
    },
    {
      "userId": 109,
      "nickname": "jaera",
      "profileImageUrl": "https://mock-bucket.s3.amazonaws.com/profile/jaera.jpg",
      "job": "기타"
    },
    {
      "userId": 101,
      "nickname": "jaemin",
      "profileImageUrl": "https://mock-bucket.s3.amazonaws.com/profile/jaemin.jpg",
      "job": "백엔드 개발자"
    }
  ]
}
```
#### 조회 결과 (jaeh로 검색)
```json
{
  "isSuccess": true,
  "code": "USER2000",
  "message": "사용자 조회 성공",
  "result": [
    {
      "userId": 102,
      "nickname": "jaehyun",
      "profileImageUrl": "https://mock-bucket.s3.amazonaws.com/profile/jaehyun.jpg",
      "job": "프론트엔드 개발자"
    },
    {
      "userId": 106,
      "nickname": "leejaehoon",
      "profileImageUrl": "https://mock-bucket.s3.amazonaws.com/profile/leejaehoon.jpg",
      "job": "학생"
    }
  ]
}
```
#### 없는 사용자일 때
```json
{
  "isSuccess": false,
  "code": "USER4040",
  "message": "존재하지 않는 사용자입니다."
}
```
#### 본인을 검색했을 때
```json
{
  "isSuccess": false,
  "code": "USER4040",
  "message": "존재하지 않는 사용자입니다."
}
```

yangjiae12 committed:
Feat: 닉네임 사용자 검색 API 엔드포인트 추가

- GET /api/users/search
- 닉네임과 userId를 파라미터로 받아 사용자 ID 반환

yangjiae12 committed:
Feat: 닉네임 검색 서비스 로직 구현

- 닉네임으로 사용자 조회 후 본인 여부 필터링
- 존재하지 않거나 자기 자신일 경우 USER_NOT_FOUND 예외 반환

yangjiae12 committed:
Feat: 닉네임 기반 사용자 조회 쿼리 추가

- findByNickname(String nickname) 메서드 정의

yangjiae12 committed:
Feat: 사용자 성공/실패 상태 코드 Enum 정의

- USER_NOT_FOUND, _USER_FOUND 상태 정의
- ApiResponse 및 예외 처리용으로 사용

Jinseong01 commented:
닉네임으로 친구 검색 (친구 userId 반환) -> [??] -> 친구 프로필 (친구 userId로 검색)
이 중간과정이 어떻게 정리되는 건 지 알 수 있을까요??
피그마에는 안 보여서 여쭤봅니다!

중간 화면(?)없이 바로 전환되는 걸까요??

yangjiae12 commented:
> 닉네임으로 친구 검색 (친구 userId 반환) -> [??] -> 친구 프로필 (친구 userId로 검색) 이 중간과정이 어떻게 정리되는 건 지 알 수 있을까요?? 피그마에는 안 보여서 여쭤봅니다!
> 
> 중간 과정없이 바로 전환되는 걸까요??

중간 과정이 따로 없어서 바로 전환으로 이해하고 구현하긴 했습니다!

> 닉네임을 검색하면 5.4의 친구 프로필 화면으로 전환하고 친구 프로필 내에서 팔로잉을 할 수 있는 것일까요?

이런 질문도 했었는데 팔로잉 버튼 추가에 대한 답변만 받고 전환에 대한 확답까진 받지 못했었네요/... 바로 친구 프로필로 전환하는 거면 그대로 유지해도 될 것 같은데 우선은 혜빈 님께 다시 질문드리겠습니다!

Jinseong01 commented:
> 이런 질문도 했었는데 팔로잉 버튼 추가에 대한 답변만 받고 전환에 대한 확답까진 받지 못했었네요/... 바로 친구 프로필로 전환하는 거면 그대로 유지해도 될 것 같은데 우선은 혜빈 님께 다시 질문드리겠습니다!

넵! 그리고 중간에 UI 없이 [친구01 - 닉네임 검색 -> 친구04 - 해당 사용자의 프로필] 식이라면 API 통합도 고려해봐도 좋을 거 같습니다!

yangjiae12 commented:
> > 이런 질문도 했었는데 팔로잉 버튼 추가에 대한 답변만 받고 전환에 대한 확답까진 받지 못했었네요/... 바로 친구 프로필로 전환하는 거면 그대로 유지해도 될 것 같은데 우선은 혜빈 님께 다시 질문드리겠습니다!
> 
> 넵! 그리고 중간에 UI 없이 [친구01 - 닉네임 검색 -> 친구04 - 해당 사용자의 프로필] 식이라면 API 통합도 고려해봐도 좋을 거 같습니다!

처음에는 닉네임이 완전히 일치하는 경우 유저아이디를 반환한다고 생각하고 검색 API를 구현했는데, 부분 일치 검색 + 여러 사용자 리스트를 반환하는 구조라고 답변 받았습니다! 기획에 맞추어서 수정한 다음에 리뷰 요청 누를테니 그때 확인하고 승인해주시면 감사하겠습니다~~

yangjiae12 committed:
Feat: 닉네임 검색 응답용 UserSearchResponseDto 생성

- userId, nickname, profileImageUrl, job 포함
- 닉네임 부분 검색 결과 반환용 DTO로 사용

yangjiae12 committed:
Chore: Levenshtein 유사도 계산을 위한 commons-text 라이브러리 추가

yangjiae12 committed:
Feat: 닉네임 검색 기능을 부분 일치 + 유사도 정렬로 변경

- findByNickname → findByNicknameContaining
- LevenshteinDistance 기반 유사도 계산 및 정렬
- 최대 5개의 결과 반환

yangjiae12 committed:
Refactor: 응답 변환 로직 분리 및 S3 이미지 URL 반환 적용

- UserConverter 생성 및 toSearchResponseDto 분리
- profileImageKey → S3Util.toUrl(profileImageKey) 로 변환
- Controller 응답 타입을 List<UserSearchResponseDto>로 변경

yangjiae12 committed:
Refactor: 닉네임 유사도 검색 로직 가독성 및 성능 개선

- keyword.toLowerCase() 반복 호출 제거
- Comparator 분리로 정렬 기준 명시
- 람다 복잡도 낮추고 가독성 향상
- 전체 흐름은 동일하며 유지보수성 강화

hyeonda02 committed:
Feat : S3 Presigned URL 방식 적용 및 관련 설정 추가
- S3Util에 toPresignedUrl(String key, Duration duration)메서드 추가
- S3Config에 S3Presigner Bean 등록

hyeonda02 committed:
A
Refactor : S3Util 프로퍼티 수정

hyeonda02 commented on src/main/java/umc/teumteum/server/domain/user/service/UserServiceImpl.java:
```diff
@@ -1,30 +1,51 @@
 @RequiredArgsConstructor
 public class UserServiceImpl implements UserService {
 
+    @Autowired
```
`UserServiceImpl`에서 `@Autowired`가 final필드에 붙어 있는데 제거해도 괜찮을 것 같습니다!
클래스에 이미 `RequiredArgsConstructor`가 선언되어 있어서 final이 붙은 필드들을 생성자를 통해 의존성을 주입해주기 때문에 `Autowired` 없어도 자동 주입 됩니당
develop 최신 반영돼서 한번 pull 받고 수정해주시면 좋을 것 같아요!

yangjiae12 replied:
넵 바로 수정하겠습니다~!

hyeonda02 replied:
UserConverter 쪽에 S3Util (toUrl -> toPresignedUrl)로 변경된거 반영해주셔야 할거 같아용 그것때문에 CI에서 빌드 실패 나는게 아닐까.. 싶은??

yangjiae12 committed:
Merge branch 'develop' of https://github.com/UMC8-TeumTeum/BE into feature/41-nickname-search

yangjiae12 committed:
Refactor: UserServiceImpl에서 final 필드의 @Autowired 제거

yangjiae12 committed:
Refactor: 사용자 프로필 이미지 URL 반환 방식 Presigned URL로 변경

hyeonda02 commented on src/main/java/umc/teumteum/server/global/util/S3Util.java:
```diff
@@ -12,7 +12,7 @@
 public class S3Util {
 
-    @Value("${cloud.aws.s3.bucket}")
+    @Value("${cloud.aws.s3.bucket.name}")
```
뭐지.. devleop에서 수정된게 안받아와진건가ㅠㅜ 분명 이부분 수정해놨었는데.. 이상하네욥
cloud.aws.s3.bucket 로 바꿔주세요!

yangjiae12 replied:
앗 안 그래도 거기서 오류 나는 거 같아서 찾아보고 있었는데 감사합니다ㅜㅜ!!!

yangjiae12 committed:
Fix: S3Util 설정 키 수정으로 주입 오류 해결

hyeonda02 reviewed (approved):
고생 많으셨습니다ㅠㅜ !! 👍🥺
